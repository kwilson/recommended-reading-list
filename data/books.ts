import { parseIsbn, parseString, parseNumber } from '../utils/data-parser';

export interface IBookDataResponseItem {
  id: string;
  cover: string;
  isbn: string;
  title: string;
  subtitle: string;
  author: string;
  published: string;
  publisher?: string;
  pages?: number;
  description?: string;
  website?: string;
}

export interface IBookDataResponse {
  books: IBookDataResponseItem[];
}

export interface IBook {
  id: string;
  title: string;
  author: string;
  cover: string;
}

export interface IBookDetails extends IBook {
  isbn: string | null;
  subtitle: string;
  published: string;
  publisher: string | null;
  pages: number | null;
  description: string | null;
  website: string | null;
}

const endpoint = 'https://hokodo-frontend-interview.netlify.app/data.json';

async function getBooksData(): Promise<IBookDataResponse> {
  const data = await fetch(endpoint);
  const values: IBookDataResponse = await data.json();

  return values;
}

export const getBooks = async (): Promise<IBook[]> => {
  try {
    const data = await getBooksData();
    return data.books.map((input) => ({
      author: input.author,
      cover: input.cover,
      id: input.id,
      title: input.title,
    }));
  } catch (e) {
    return [];
  }
};

export const getBooksByAuthor = async (
  author: string,
  exludeIds: string[] = []
): Promise<IBook[]> => {
  try {
    const data = await getBooksData();
    const predicate = (book: IBook) =>
      book.author === author && !exludeIds.includes(book.id);

    return data.books.filter(predicate).map((input) => ({
      author: input.author,
      cover: input.cover,
      id: input.id,
      title: input.title,
    }));
  } catch (e) {
    return [];
  }
};

export const getBook = async (
  id: string | null
): Promise<IBookDetails | null> => {
  try {
    const { books } = await getBooksData();
    const match = books.find((book) => book.id === id);

    if (match) {
      return {
        author: match.author,
        cover: match.cover,
        description: parseString(match.description),
        id: match.id,
        isbn: parseIsbn(match.isbn),
        pages: parseNumber(match.pages),
        published: match.published,
        publisher: parseString(match.publisher),
        subtitle: match.subtitle,
        title: match.title,
        website: parseString(match.website),
      };
    }

    return null;
  } catch (e) {
    return null;
  }
};
