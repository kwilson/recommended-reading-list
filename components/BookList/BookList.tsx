import { FunctionComponent } from 'react';

import { BookListItem } from '../BookListItem';
import { IBook } from '../../data/books';

import styles from './BookList.module.css';

interface IProps {
  books: IBook[];
}

export const BookList: FunctionComponent<IProps> = ({ books = [] }) => {
  return (
    <div className={styles.container}>
      {books.length === 0 && (
        <div data-testid="empty-book-list">
          <p>Sorry, there are no books to display.</p>
        </div>
      )}

      {books.length > 0 && (
        <div data-testid="book-list-items">
          {books.map(({ id, author, title, cover }) => (
            <BookListItem
              key={id}
              author={author}
              cover={cover}
              title={title}
              link={`/${id}`}
            />
          ))}
        </div>
      )}
    </div>
  );
};
