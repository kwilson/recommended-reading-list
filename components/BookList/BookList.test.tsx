import { render, screen } from '@testing-library/react';

import { IBook } from '../../data/books';
import { BookList } from './BookList';

describe(`<${BookList.name} />`, () => {
  it('renders as expected when there are no books', async () => {
    const { asFragment } = render(<BookList books={[]} />);

    expect(screen.getByTestId('empty-book-list')).toBeDefined();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders as expected when there are books', async () => {
    const testData: IBook[] = [1, 2, 3, 4, 5, 6, 7, 8, 9].map((key) => ({
      id: `id-${key}`,
      cover: `url://${key}`,
      title: `title-${key}`,
      author: `author-${key % 2}`,
    }));

    const { asFragment } = render(<BookList books={testData} />);

    expect(screen.queryByTestId('empty-book-list')).toBe(null);

    testData.forEach((book) => {
      expect(screen.getByText(book.title)).toBeDefined();
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
