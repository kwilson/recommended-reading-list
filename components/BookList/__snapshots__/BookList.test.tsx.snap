// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BookList /> renders as expected when there are books 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      data-testid="book-list-items"
    >
      <div
        class="container"
      >
        <div
          class="details"
        >
          <h2
            class="title"
            data-testid="book-title"
          >
            title-1
          </h2>
          <div
            class="author"
          >
            by 
            <span
              data-testid="book-author"
            >
              author-1
            </span>
          </div>
          <div
            class="link"
            data-testid="book-link"
          >
            <a
              href="/id-1"
            >
              More Info
            </a>
          </div>
        </div>
        <div
          class="cover"
        >
          <img
            alt="title-1 book cover"
            class="coverImage"
            data-testid="book-image"
            src="url://1"
          />
        </div>
      </div>
      <div
        class="container"
      >
        <div
          class="details"
        >
          <h2
            class="title"
            data-testid="book-title"
          >
            title-2
          </h2>
          <div
            class="author"
          >
            by 
            <span
              data-testid="book-author"
            >
              author-0
            </span>
          </div>
          <div
            class="link"
            data-testid="book-link"
          >
            <a
              href="/id-2"
            >
              More Info
            </a>
          </div>
        </div>
        <div
          class="cover"
        >
          <img
            alt="title-2 book cover"
            class="coverImage"
            data-testid="book-image"
            src="url://2"
          />
        </div>
      </div>
      <div
        class="container"
      >
        <div
          class="details"
        >
          <h2
            class="title"
            data-testid="book-title"
          >
            title-3
          </h2>
          <div
            class="author"
          >
            by 
            <span
              data-testid="book-author"
            >
              author-1
            </span>
          </div>
          <div
            class="link"
            data-testid="book-link"
          >
            <a
              href="/id-3"
            >
              More Info
            </a>
          </div>
        </div>
        <div
          class="cover"
        >
          <img
            alt="title-3 book cover"
            class="coverImage"
            data-testid="book-image"
            src="url://3"
          />
        </div>
      </div>
      <div
        class="container"
      >
        <div
          class="details"
        >
          <h2
            class="title"
            data-testid="book-title"
          >
            title-4
          </h2>
          <div
            class="author"
          >
            by 
            <span
              data-testid="book-author"
            >
              author-0
            </span>
          </div>
          <div
            class="link"
            data-testid="book-link"
          >
            <a
              href="/id-4"
            >
              More Info
            </a>
          </div>
        </div>
        <div
          class="cover"
        >
          <img
            alt="title-4 book cover"
            class="coverImage"
            data-testid="book-image"
            src="url://4"
          />
        </div>
      </div>
      <div
        class="container"
      >
        <div
          class="details"
        >
          <h2
            class="title"
            data-testid="book-title"
          >
            title-5
          </h2>
          <div
            class="author"
          >
            by 
            <span
              data-testid="book-author"
            >
              author-1
            </span>
          </div>
          <div
            class="link"
            data-testid="book-link"
          >
            <a
              href="/id-5"
            >
              More Info
            </a>
          </div>
        </div>
        <div
          class="cover"
        >
          <img
            alt="title-5 book cover"
            class="coverImage"
            data-testid="book-image"
            src="url://5"
          />
        </div>
      </div>
      <div
        class="container"
      >
        <div
          class="details"
        >
          <h2
            class="title"
            data-testid="book-title"
          >
            title-6
          </h2>
          <div
            class="author"
          >
            by 
            <span
              data-testid="book-author"
            >
              author-0
            </span>
          </div>
          <div
            class="link"
            data-testid="book-link"
          >
            <a
              href="/id-6"
            >
              More Info
            </a>
          </div>
        </div>
        <div
          class="cover"
        >
          <img
            alt="title-6 book cover"
            class="coverImage"
            data-testid="book-image"
            src="url://6"
          />
        </div>
      </div>
      <div
        class="container"
      >
        <div
          class="details"
        >
          <h2
            class="title"
            data-testid="book-title"
          >
            title-7
          </h2>
          <div
            class="author"
          >
            by 
            <span
              data-testid="book-author"
            >
              author-1
            </span>
          </div>
          <div
            class="link"
            data-testid="book-link"
          >
            <a
              href="/id-7"
            >
              More Info
            </a>
          </div>
        </div>
        <div
          class="cover"
        >
          <img
            alt="title-7 book cover"
            class="coverImage"
            data-testid="book-image"
            src="url://7"
          />
        </div>
      </div>
      <div
        class="container"
      >
        <div
          class="details"
        >
          <h2
            class="title"
            data-testid="book-title"
          >
            title-8
          </h2>
          <div
            class="author"
          >
            by 
            <span
              data-testid="book-author"
            >
              author-0
            </span>
          </div>
          <div
            class="link"
            data-testid="book-link"
          >
            <a
              href="/id-8"
            >
              More Info
            </a>
          </div>
        </div>
        <div
          class="cover"
        >
          <img
            alt="title-8 book cover"
            class="coverImage"
            data-testid="book-image"
            src="url://8"
          />
        </div>
      </div>
      <div
        class="container"
      >
        <div
          class="details"
        >
          <h2
            class="title"
            data-testid="book-title"
          >
            title-9
          </h2>
          <div
            class="author"
          >
            by 
            <span
              data-testid="book-author"
            >
              author-1
            </span>
          </div>
          <div
            class="link"
            data-testid="book-link"
          >
            <a
              href="/id-9"
            >
              More Info
            </a>
          </div>
        </div>
        <div
          class="cover"
        >
          <img
            alt="title-9 book cover"
            class="coverImage"
            data-testid="book-image"
            src="url://9"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<BookList /> renders as expected when there are no books 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      data-testid="empty-book-list"
    >
      <p>
        Sorry, there are no books to display.
      </p>
    </div>
  </div>
</DocumentFragment>
`;
